torch/_dynamo/guards.py

```diff
@@ -414,9 +418,13 @@ def TENSOR_MATCH(self, guard: Guard):
             self.ID_MATCH(guard)
         else:
             value = self.get(guard.name)
-            self.tensor_check_names.append(self.arg_ref(guard))
+            tensor_name = self.arg_ref(guard)
+            self.tensor_check_names.append(tensor_name)
             self.tensor_check_examples.append(value)
 
+            # STOP - DO NOT USE id_ref FOR TENSORS - TENSOR INVALIDATION RULES DIFFER
```

Comment:
I don't know what this means

Reply:
This is an uncommon pattern of extracting an id out when installing guards. Calling id_ref is the norm, as it goes through an  invalidation registration but we specifically need to avoid that for tensors.

Reply:
putting on my noob hat for a moment: what are the "tensor invalidation rules"? What's an "invalidation registratoin"? How do I decide if I call `id_ref` or the function you call here? (Pointers to docs for these OK.)

Reply:
Good question! 

id references in guards are brittle - because the id may go out of scope. Using it may then give us garbage memory (as id is an address, as you know), or worse (segfaults, crashes, etc).

So, in general, when you use an `id` that will make it up into the guard code, you need a way of protecting that (lest we use an old id). So, we have a registration hook:

```
    def id_ref(self, obj):
        """add a weakref, return the id"""
        try:
            if id(obj) not in self._seen_ids:
                self._weakrefs.append(weakref.ref(obj, self.invalidate))
                self._seen_ids.add(id(obj))
        except TypeError:
            pass  # cannot weakref bool object
        return id(obj)
```

This little wrapper registers the id with an invalidation callback, which, upon weakref invalidation, flips the whole guard's `check_fn` to "invalid", essentially meaning that any future invocations of this `check_fn` to return false, aka a cache miss, as `.valid` is always checked first (for this safety reason).

The ID here will CATEGORICALLY NOT go into a check_fn (for reasons you stated elsewhere in this review). Instead, its used for intermediary lookup between tensor name and shape expression. So, registering it for invalidation is the wrong thing to do, as we DO NOT want to couple that tensor going out of scope and getting invalidated (fine) with invalidating the entire `check_fn` - because the `check_fn` does not reference the id, it references the tensor's **name**.

Reply:
Using a weakref is fine. Registration of the invalidation of said weakref with the rest of the guard fn is not. Do you want me to make this a weakref?

Reply:
I'm not asking for anything, I'm just trying to figure out why the more obvious thing was not done. In this case, I think I'm reasonably sure the optimization is profitable, and so I don't want you to make it use a weakref, I want to confirm that I diagnosed it correctly

Reply:
Yes :)